File: components/credentials/src/main/java/org/cloudfoundry/credhub/data/ExpiryDateMigration.java
package org.cloudfoundry.credhub.data;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.cloudfoundry.credhub.entity.CertificateCredentialVersionData;
import org.cloudfoundry.credhub.entity.CredentialVersionData;
import org.cloudfoundry.credhub.exceptions.MalformedCertificateException;
import org.cloudfoundry.credhub.repositories.CredentialVersionRepository;
import org.cloudfoundry.credhub.utils.CertificateReader;

@Component
public class ExpiryDateMigration {
  private static final Logger LOGGER = LogManager.getLogger(ExpiryDateMigration.class);

  private final CredentialVersionRepository credentialVersionRepository;

  @Autowired
  public ExpiryDateMigration(final CredentialVersionRepository credentialVersionRepository) {
    super();
    this.credentialVersionRepository = credentialVersionRepository;
  }

  public void migrate() {
    final List<CredentialVersionData> data = credentialVersionRepository.findAllVersionsWithNullExpirationDate();

    for (final CredentialVersionData version : data) {
      if (version instanceof CertificateCredentialVersionData) {
        final CertificateCredentialVersionData certificateVersion = (CertificateCredentialVersionData) version;
        final String certificate = certificateVersion.getCertificate();
        try {
          final CertificateReader reader = new CertificateReader(certificate);
          ((CertificateCredentialVersionData) version).setExpiryDate(reader.getNotAfter());
        } catch (MalformedCertificateException e) {
          final String message = String.format("can't read certificate with name %s", certificateVersion.getName());
          LOGGER.warn(message);
        }
      }
    }
    credentialVersionRepository.saveAll(data);
  }
}

File: components/credentials/src/main/kotlin/org/cloudfoundry/credhub/data/ExpiryDateMigration.kt
package org.cloudfoundry.credhub.data

import org.apache.logging.log4j.LogManager
import org.cloudfoundry.credhub.entity.CertificateCredentialVersionData
import org.cloudfoundry.credhub.exceptions.MalformedCertificateException
import org.cloudfoundry.credhub.exceptions.MissingCertificateException
import org.cloudfoundry.credhub.repositories.CredentialVersionRepository
import org.cloudfoundry.credhub.utils.CertificateReader
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class ExpiryDateMigration @Autowired
constructor(private val credentialVersionRepository: CredentialVersionRepository) {
    fun migrate() {
        var numberOfInvalidCertsFound = 0
        while (credentialVersionRepository.countVersionsWithNullExpirationDate() != numberOfInvalidCertsFound) {
            val data = credentialVersionRepository.findUpTo1000VersionsWithNullExpirationDate(numberOfInvalidCertsFound)
            for (version in data) {
                if (version is CertificateCredentialVersionData) {
                    val certificate = version.certificate
                    try {
                        val reader = CertificateReader(certificate)
                        version.expiryDate = reader.notAfter
                    } catch (e: RuntimeException) {
                        printErrorMessage(e, version)
                        numberOfInvalidCertsFound++
                    }
                }
            }
            credentialVersionRepository.saveAll(data)
        }
    }

    fun printErrorMessage(e: RuntimeException, version: CertificateCredentialVersionData) {
        var message = String.format("Unexpected exception reading certificate with name %s: %s", version.name, e.toString())
        when (e) {
            is MalformedCertificateException ->
                message = String.format("can't read certificate with name %s", version.getName())
            is MissingCertificateException ->
                message = String.format("missing certificate with name %s", version.name)
        }
        LOGGER.warn(message)
    }

    companion object {
        private val LOGGER = LogManager.getLogger(ExpiryDateMigration::class.java)
    }
}
